File: typescript/src/models/str-id.ts
import slug from 'slug';
import { v4 as uuid } from 'uuid';

export const ErrInvalidId = new Error('invalid id');

export class StrId {
  private id: string;

  constructor(id: string) {
    if (id.length == 0) {
      throw ErrInvalidId;
    }

    this.id = id;
  }

  static generateUuid(): StrId {
    return new StrId(uuid());
  }

  static generateSlug(str: string): StrId {
    return new StrId(slug(str));
  }

  value(): string {
    return this.id;
  }

  equals(other: StrId): boolean {
    return this.id === other.id;
  }
}

File: go/models/str_id.go
package models

import (
	"github.com/google/uuid"
	"github.com/gosimple/slug"
)

type StrId struct {
	id string
}

func NewStrId(id string) (StrId, error) {
	if len(id) == 0 {
		return StrId{}, &ErrInvalidId{id}
	}

	return StrId{
		id: id,
	}, nil
}

func GenerateUuid() StrId {
	id, _ := NewStrId(uuid.New().String())
	return id
}

func GenerateSlug(str string) (StrId, error) {
	s := slug.Make(str)
	return NewStrId(s)
}

func (id StrId) Value() string {
	return id.id
}

func (id StrId) Equals(other StrId) bool {
	return id.id == other.id
}
